File: FastMines_android/src/main/java/fmg/android/app/model/items/MosaicSkillDataItem.java
package fmg.android.app.model.items;

import androidx.databinding.Bindable;

import java.util.concurrent.ThreadLocalRandom;

import fmg.android.app.BR;
import fmg.android.img.AndroidBitmapView;
import fmg.android.img.MosaicSkillImg;
import fmg.common.geom.BoundDouble;
import fmg.core.img.MosaicSkillModel;
import fmg.core.types.ESkillLevel;

import static fmg.core.img.PropertyConst.PROPERTY_MODEL_BURGER_MENU_PADDING;
import static fmg.core.img.PropertyConst.PROPERTY_SKILL_LEVEL;

/** Mosaic skill level item for data model */
public class MosaicSkillDataItem extends BaseDataItem<ESkillLevel, MosaicSkillModel, AndroidBitmapView<MosaicSkillModel>, MosaicSkillImg.MosaicSkillAndroidBitmapController> {

    public MosaicSkillDataItem(ESkillLevel eSkill) {
        super(eSkill);
        setTitle(eSkill==null ? null : eSkill.getDescription());
    }

    @Bindable
    public ESkillLevel getSkillLevel()                   { return getUniqueId(); }
    public void        setSkillLevel(ESkillLevel eSkill) {        setUniqueId(eSkill); }

    @Deprecated
    public String getUnicodeChar() { return getSkillLevel() == null ? null : Character.toString(getSkillLevel().unicodeChar()); }

    @Override
    public MosaicSkillImg.MosaicSkillAndroidBitmapController getController() {
        if (this.controller == null) {
            var ctrl = new MosaicSkillImg.MosaicSkillAndroidBitmapController(getSkillLevel());
            var m = ctrl.getModel();
            m.setBorderWidth(2);
            m.setRotateAngle(ThreadLocalRandom.current().nextInt(90));
            setController(ctrl);
        }
        return this.controller;
    }

    @Bindable
    public BoundDouble getPaddingBurgerMenu() {
        BoundDouble pad = getController().getBurgerMenuModel().getPadding();
        double zoom = getZoom();
        return new BoundDouble(pad.left / zoom, pad.top / zoom, pad.right / zoom, pad.bottom / zoom);
    }

    public void setPaddingBurgerMenu(BoundDouble paddingBurgerMenu) {
        getController().getBurgerMenuModel().setPadding(zoomPadding(paddingBurgerMenu));
    }

    @Bindable
    public String getContentDescription() {
        return "MosaicSkillItem:" + getSkillLevel().getDescription();
    }

    @Override
    protected void firePropertyChanged(String propertyName) {
        if (isDisposed())
            return;

        super.firePropertyChanged(propertyName);

        switch (propertyName) {
        case PROPERTY_UNIQUE_ID:
            getController().getModel().setMosaicSkill(getSkillLevel());
            break;
        case PROPERTY_SKILL_LEVEL:
            super.firePropertyChanged(PROPERTY_UNIQUE_ID);
            break;
        }
    }

    @Override
    protected void firePropertyChangedAsync(String propertyName) {
        if (isDisposed())
            return;

        super.firePropertyChangedAsync(propertyName);

        // refire as android data binding event
        switch (propertyName) {
        case PROPERTY_SKILL_LEVEL              : notifyPropertyChanged(BR.skillLevel       ); break;
        case PROPERTY_MODEL_BURGER_MENU_PADDING: notifyPropertyChanged(BR.paddingBurgerMenu); break;
        }
    }

    @Override
    public void close() {
        getController().getBurgerMenuModel().setListener(null);
        super.close();
    }

}

File: FastMines_uwp/src/fmg/uwp/app/Model/Items/MosaicSkillDataItem.cs
using System;
using System.ComponentModel;
using Microsoft.Graphics.Canvas;
using Fmg.Common;
using Fmg.Common.Notifier;
using Fmg.Common.Geom;
using Fmg.Core.Types;
using Fmg.Core.Img;
using MosaicSkillView       = Fmg.Uwp.Img.Win2d.MosaicSkillImg.CanvasBmpView;
using MosaicSkillController = Fmg.Uwp.Img.Win2d.MosaicSkillImg.CanvasBmpController;

namespace Fmg.Uwp.App.Model.Items {

    /// <summary> Mosaic skill level item for data model </summary>
    public class MosaicSkillDataItem : BaseDataItem<ESkillLevel?, MosaicSkillModel, MosaicSkillView, MosaicSkillController> {

        public MosaicSkillDataItem(ESkillLevel? eSkill)
            : base(eSkill)
        {
            Title = eSkill?.GetDescription();
        }

        public ESkillLevel? SkillLevel => UniqueId;

        [Obsolete]
        public string UnicodeChar => SkillLevel?.UnicodeChar().ToString();

        public override MosaicSkillController Entity {
            get {
                if (entity == null) {
                    var tmp = new MosaicSkillController(SkillLevel, CanvasDevice.GetSharedDevice());
                    var m = tmp.Model;
                    m.BorderWidth = 2;
                    m.RotateAngle = ThreadLocalRandom.Current.Next(90);
                    tmp.BurgerMenuModel.PropertyChanged += OnBurgerMenuModelPropertyChanged;
                    Entity = tmp; // call this setter
                }
                return entity;
            }
        }

        public BoundDouble PaddingBurgerMenu {
            get {
                var pad = Entity.BurgerMenuModel.Padding;
                var zoom = Zoom();
                return new BoundDouble(pad.Left / zoom, pad.Top / zoom, pad.Right / zoom, pad.Bottom / zoom);
            }
            set {
                Entity.BurgerMenuModel.Padding = ZoomPadding(value);
            }
        }

        protected override void OnPropertyChanged(object sender, PropertyChangedEventArgs ev) {
            base.OnPropertyChanged(sender, ev);

            switch(ev.PropertyName) {
            case nameof(this.UniqueId):
                _notifier.FirePropertyChanged(nameof(this.SkillLevel)); // recall with another property name
                break;
            }
        }

        protected void OnBurgerMenuModelPropertyChanged(object sender, PropertyChangedEventArgs ev) {
            switch (ev.PropertyName) {
            case nameof(BurgerMenuModel.Padding):
                if (ev is PropertyChangedExEventArgs<BoundDouble> evx)
                    _notifier.FirePropertyChanged(ZoomPadding(evx.OldValue), ZoomPadding(evx.NewValue), nameof(this.PaddingBurgerMenu));
                else
                    _notifier.FirePropertyChanged(nameof(this.PaddingBurgerMenu));
                break;
            }
        }

        protected override void Disposing() {
            Entity.BurgerMenuModel.PropertyChanged -= OnBurgerMenuModelPropertyChanged;
            base.Disposing();
        }

    }

}
